unit/util/piped_process.cpp

```diff
@@ -0,0 +1,98 @@
+/// \file
+/// \author Diffblue Ltd.
+/// Unit tests for checking the piped process communication mechanism.
+
+#ifdef _WIN32
+// No unit tests yet!
+#else
+
+#  include <testing-utils/use_catch.h>
+#  include <unistd.h>
+#  include <util/piped_process.h>
+
+TEST_CASE(
+  "We create a pipe and we can read from it",
+  "[core][util][piped_process]")
+{
+  std::string to_be_echoed = "The Jabberwocky";
+  // Need to give up to avoid shell built-in invocation
+  std::string binary = "/bin/echo";
+  std::string command = binary + " " + to_be_echoed;
+  piped_processt process = piped_processt(command);
+
+  sleep(1);
+
+  std::string response = process.receive();
+
+  // Trim newline from response string that causes match to fail.
+  response.erase(
+    std::remove(response.begin(), response.end(), '\n'), response.end());
+
+  REQUIRE(response == to_be_echoed);
+}
+
+TEST_CASE(
+  "We create a pipe, send and receive from it",
+  "[core][util][piped_process]")
+{
+  std::string binary = "z3 -in";
+  std::string statement = "(echo \"hi\")";
+  std::string termination_statement = "(exit)";
+  piped_processt process = piped_processt(binary);
+  // Wait a moment for z3 to start
```

Comment:
Eh?  Why?  Pipes should be reliable and asynchronous.

Reply:
Pipes are asynchronous, but the pipes here are non-blocking. This means if you try to read immediately the other process may have not written anything to the pipe yet and you can read nothing (and then fail the following test).

Having the ability to read from a pipe that is not ended/closed is significant to being able to interact (e.g. with a solver) without requiring it to be terminated/closed after each request for a response. That said, without an explicit synchronisation system (yet?) it will not always be clear when the pipe is ready to be read. An alternate is to read lines, but this requires knowing how many lines to read, which may not be known in advance (e.g. how many lines make up a trace?) or doing parsing/checking after each line to know when to continue/stop. (This PR is clearly WIP, mostly I believe created to test the CI and various builds, it is not ready for review or merging.)